File: apps/mobile/android/app/src/main/java/com/tchat/components/TchatCard.kt
package com.tchat.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.tchat.designsystem.Colors
import com.tchat.designsystem.Spacing

/**
 * Card component following Tchat design system
 */
@Composable
fun TchatCard(
    modifier: Modifier = Modifier,
    variant: TchatCardVariant = TchatCardVariant.Elevated,
    size: TchatCardSize = TchatCardSize.Standard,
    isInteractive: Boolean = false,
    onClick: (() -> Unit)? = null,
    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
    content: @Composable ColumnScope.() -> Unit
) {
    val isPressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(
        targetValue = if (isInteractive && isPressed) 0.98f else 1f,
        label = "card_scale"
    )

    val shape = RoundedCornerShape(size.cornerRadius)

    Card(
        modifier = modifier
            .scale(scale)
            .then(
                if (isInteractive && onClick != null) {
                    Modifier.clickable(
                        interactionSource = interactionSource,
                        indication = null
                    ) { onClick() }
                } else Modifier
            ),
        shape = shape,
        colors = CardDefaults.cardColors(
            containerColor = variant.backgroundColor
        ),
        elevation = CardDefaults.cardElevation(
            defaultElevation = variant.elevation
        ),
        border = variant.borderStroke
    ) {
        Column(
            modifier = Modifier.padding(size.contentPadding),
            content = content
        )
    }
}

/**
 * Card header component
 */
@Composable
fun TchatCardHeader(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    leadingIcon: ImageVector? = null,
    trailingContent: (@Composable () -> Unit)? = null
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.sm)
    ) {
        if (leadingIcon != null) {
            Icon(
                imageVector = leadingIcon,
                contentDescription = null,
                tint = Colors.primary,
                modifier = Modifier.size(20.dp)
            )
        }

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(Spacing.xs)
        ) {
            Text(
                text = title,
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
                color = Colors.textPrimary
            )

            if (subtitle != null) {
                Text(
                    text = subtitle,
                    fontSize = 12.sp,
                    color = Colors.textSecondary
                )
            }
        }

        trailingContent?.invoke()
    }
}

/**
 * Card footer component
 */
@Composable
fun TchatCardFooter(
    modifier: Modifier = Modifier,
    content: @Composable RowScope.() -> Unit
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = Spacing.sm),
        content = content
    )
}

/**
 * Card variant definitions
 */
enum class TchatCardVariant(
    val backgroundColor: Color,
    val elevation: androidx.compose.ui.unit.Dp,
    val borderStroke: BorderStroke? = null
) {
    Elevated(
        backgroundColor = Colors.cardBackground,
        elevation = 4.dp
    ),
    Outlined(
        backgroundColor = Colors.cardBackground,
        elevation = 0.dp,
        borderStroke = BorderStroke(1.dp, Colors.border)
    ),
    Filled(
        backgroundColor = Colors.surface,
        elevation = 0.dp
    ),
    Glass(
        backgroundColor = Colors.cardBackground.copy(alpha = 0.8f),
        elevation = 2.dp
    )
}

/**
 * Card size definitions
 */
enum class TchatCardSize(
    val contentPadding: androidx.compose.foundation.layout.PaddingValues,
    val cornerRadius: androidx.compose.ui.unit.Dp
) {
    Compact(
        contentPadding = PaddingValues(Spacing.sm),
        cornerRadius = Spacing.sm
    ),
    Standard(
        contentPadding = PaddingValues(Spacing.md),
        cornerRadius = Spacing.md
    ),
    Expanded(
        contentPadding = PaddingValues(Spacing.lg),
        cornerRadius = Spacing.md
    )
}

/**
 * Convenience composable for card with header
 */
@Composable
fun TchatCard(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    leadingIcon: ImageVector? = null,
    trailingContent: (@Composable () -> Unit)? = null,
    variant: TchatCardVariant = TchatCardVariant.Elevated,
    size: TchatCardSize = TchatCardSize.Standard,
    isInteractive: Boolean = false,
    onClick: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit
) {
    TchatCard(
        modifier = modifier,
        variant = variant,
        size = size,
        isInteractive = isInteractive,
        onClick = onClick
    ) {
        TchatCardHeader(
            title = title,
            subtitle = subtitle,
            leadingIcon = leadingIcon,
            trailingContent = trailingContent
        )

        Spacer(modifier = Modifier.height(Spacing.sm))

        content()
    }
}

// Preview
@Preview(showBackground = true)
@Composable
fun TchatCardPreview() {
    Column(
        modifier = Modifier.padding(Spacing.md),
        verticalArrangement = Arrangement.spacedBy(Spacing.md)
    ) {
        // Basic elevated card
        TchatCard(variant = TchatCardVariant.Elevated) {
            Column(verticalArrangement = Arrangement.spacedBy(Spacing.sm)) {
                Text(
                    text = "Elevated Card",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.SemiBold
                )
                Text(
                    text = "This is an elevated card with shadow",
                    color = Colors.textSecondary
                )
            }
        }

        // Outlined card with header
        TchatCard(
            title = "Card with Header",
            subtitle = "Subtitle text",
            leadingIcon = Icons.Default.Star,
            variant = TchatCardVariant.Outlined,
            isInteractive = true,
            onClick = { println("Card tapped") },
            trailingContent = {
                Icon(
                    imageVector = Icons.Default.ChevronRight,
                    contentDescription = null,
                    tint = Colors.textTertiary
                )
            }
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
                Text("This card has a header and is interactive")
                Text(
                    text = "Tap to interact",
                    fontSize = 12.sp,
                    color = Colors.primary
                )
            }
        }

        // Filled card
        TchatCard(
            variant = TchatCardVariant.Filled,
            size = TchatCardSize.Compact
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Spacing.sm)
            ) {
                Icon(
                    imageVector = Icons.Default.Message,
                    contentDescription = null,
                    tint = Colors.primary
                )
                Text(
                    text = "Compact filled card",
                    fontSize = 14.sp
                )
            }
        }

        // Glass card
        TchatCard(
            title = "Glass Card",
            subtitle = "Expanded size",
            leadingIcon = Icons.Default.AutoAwesome,
            variant = TchatCardVariant.Glass,
            size = TchatCardSize.Expanded
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(Spacing.md)) {
                Text("This is a glass-style card with expanded padding")

                TchatCardFooter {
                    TchatButton(
                        text = "Action",
                        onClick = { println("Action tapped") },
                        variant = TchatButtonVariant.Primary,
                        size = TchatButtonSize.Small
                    )

                    Spacer(modifier = Modifier.weight(1f))

                    Text(
                        text = "Footer content",
                        fontSize = 12.sp,
                        color = Colors.textTertiary
                    )
                }
            }
        }

        // Interactive card with complex content
        TchatCard(
            variant = TchatCardVariant.Elevated,
            isInteractive = true,
            onClick = { println("Complex card tapped") }
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(Spacing.md)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(
                            text = "Interactive Card",
                            fontSize = 18.sp,
                            fontWeight = FontWeight.SemiBold
                        )
                        Text(
                            text = "With complex content",
                            fontSize = 12.sp,
                            color = Colors.textSecondary
                        )
                    }
                    Icon(
                        imageVector = Icons.Default.ChevronRight,
                        contentDescription = null,
                        tint = Colors.textTertiary
                    )
                }

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.CheckCircle,
                            contentDescription = null,
                            tint = Colors.success,
                            modifier = Modifier.size(16.dp)
                        )
                        Text(
                            text = "Feature 1",
                            fontSize = 12.sp,
                            color = Colors.success
                        )
                    }

                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.Star,
                            contentDescription = null,
                            tint = Colors.warning,
                            modifier = Modifier.size(16.dp)
                        )
                        Text(
                            text = "Feature 2",
                            fontSize = 12.sp,
                            color = Colors.warning
                        )
                    }
                }
            }
        }
    }
}

File: apps/mobile/ios/Sources/Components/TchatCard.swift
//
//  TchatCard.swift
//  TchatApp
//
//  Created by Claude on 21/09/2024.
//

import SwiftUI

/// Card component following Tchat design system
public struct TchatCard<Content: View>: View {

    // MARK: - Card Variants
    public enum Variant {
        case elevated
        case outlined
        case filled
        case glass
    }

    public enum Size {
        case compact
        case standard
        case expanded
    }

    // MARK: - Properties
    let content: Content
    let variant: Variant
    let size: Size
    let isInteractive: Bool
    let onTap: (() -> Void)?

    // MARK: - Private Properties
    private let colors = Colors()
    @State private var isPressed = false

    // MARK: - Initializer
    public init(
        variant: Variant = .elevated,
        size: Size = .standard,
        isInteractive: Bool = false,
        onTap: (() -> Void)? = nil,
        @ViewBuilder content: () -> Content
    ) {
        self.content = content()
        self.variant = variant
        self.size = size
        self.isInteractive = isInteractive
        self.onTap = onTap
    }

    // MARK: - Body
    public var body: some View {
        Button(action: {
            onTap?()
        }) {
            content
                .padding(contentPadding)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(backgroundColor)
                .overlay(borderOverlay)
                .cornerRadius(cornerRadius)
                .shadow(
                    color: shadowColor,
                    radius: shadowRadius,
                    x: shadowOffset.x,
                    y: shadowOffset.y
                )
        }
        .buttonStyle(TchatCardButtonStyle(
            isInteractive: isInteractive,
            variant: variant
        ))
        .disabled(!isInteractive)
    }

    // MARK: - Computed Properties

    private var contentPadding: EdgeInsets {
        switch size {
        case .compact:
            return EdgeInsets(
                top: Spacing.sm,
                leading: Spacing.sm,
                bottom: Spacing.sm,
                trailing: Spacing.sm
            )
        case .standard:
            return EdgeInsets(
                top: Spacing.md,
                leading: Spacing.md,
                bottom: Spacing.md,
                trailing: Spacing.md
            )
        case .expanded:
            return EdgeInsets(
                top: Spacing.lg,
                leading: Spacing.lg,
                bottom: Spacing.lg,
                trailing: Spacing.lg
            )
        }
    }

    private var backgroundColor: Color {
        switch variant {
        case .elevated, .outlined:
            return colors.cardBackground
        case .filled:
            return colors.surface
        case .glass:
            return colors.cardBackground.opacity(0.8)
        }
    }

    @ViewBuilder
    private var borderOverlay: some View {
        switch variant {
        case .outlined:
            RoundedRectangle(cornerRadius: cornerRadius)
                .stroke(colors.border, lineWidth: 1)
        default:
            EmptyView()
        }
    }

    private var cornerRadius: CGFloat {
        switch size {
        case .compact:
            return Spacing.sm
        case .standard:
            return Spacing.md
        case .expanded:
            return Spacing.md
        }
    }

    private var shadowColor: Color {
        switch variant {
        case .elevated:
            return colors.shadowMedium
        case .glass:
            return colors.shadowLight
        default:
            return Color.clear
        }
    }

    private var shadowRadius: CGFloat {
        switch variant {
        case .elevated:
            return 8
        case .glass:
            return 4
        default:
            return 0
        }
    }

    private var shadowOffset: CGSize {
        switch variant {
        case .elevated:
            return CGSize(width: 0, height: 2)
        case .glass:
            return CGSize(width: 0, height: 1)
        default:
            return CGSize.zero
        }
    }
}

// MARK: - Card Button Style
struct TchatCardButtonStyle: ButtonStyle {
    let isInteractive: Bool
    let variant: TchatCard<AnyView>.Variant
    private let colors = Colors()

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(
                isInteractive && configuration.isPressed ? 0.98 : 1.0
            )
            .brightness(
                isInteractive && configuration.isPressed ? (
                    variant == .filled ? 0.05 : -0.02
                ) : 0
            )
            .animation(.easeInOut(duration: 0.1), value: configuration.isPressed)
    }
}

// MARK: - Card Header Component
public struct TchatCardHeader: View {
    let title: String
    let subtitle: String?
    let leadingIcon: String?
    let trailingContent: AnyView?

    private let colors = Colors()

    public init(
        title: String,
        subtitle: String? = nil,
        leadingIcon: String? = nil,
        trailingContent: (() -> AnyView)? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.leadingIcon = leadingIcon
        self.trailingContent = trailingContent?()
    }

    public var body: some View {
        HStack(spacing: Spacing.sm) {
            if let leadingIcon = leadingIcon {
                Image(systemName: leadingIcon)
                    .foregroundColor(colors.primary)
                    .frame(width: 20, height: 20)
            }

            VStack(alignment: .leading, spacing: Spacing.xs) {
                Text(title)
                    .font(.headline)
                    .foregroundColor(colors.textPrimary)

                if let subtitle = subtitle {
                    Text(subtitle)
                        .font(.caption)
                        .foregroundColor(colors.textSecondary)
                }
            }

            Spacer()

            trailingContent
        }
    }
}

// MARK: - Card Footer Component
public struct TchatCardFooter: View {
    let content: AnyView

    public init(@ViewBuilder content: () -> AnyView) {
        self.content = content()
    }

    public var body: some View {
        HStack {
            content
        }
        .padding(.top, Spacing.sm)
    }
}

// MARK: - Convenience Initializers
extension TchatCard {
    /// Card with header and content
    public init(
        title: String,
        subtitle: String? = nil,
        leadingIcon: String? = nil,
        variant: Variant = .elevated,
        size: Size = .standard,
        isInteractive: Bool = false,
        onTap: (() -> Void)? = nil,
        @ViewBuilder content: () -> Content
    ) where Content == AnyView {
        self.init(
            variant: variant,
            size: size,
            isInteractive: isInteractive,
            onTap: onTap
        ) {
            AnyView(
                VStack(alignment: .leading, spacing: Spacing.sm) {
                    TchatCardHeader(
                        title: title,
                        subtitle: subtitle,
                        leadingIcon: leadingIcon
                    )

                    content()
                }
            )
        }
    }
}

// MARK: - Preview
#if DEBUG
struct TchatCard_Previews: PreviewProvider {
    static var previews: some View {
        ScrollView {
            VStack(spacing: Spacing.md) {
                // Basic elevated card
                TchatCard(variant: .elevated) {
                    VStack(alignment: .leading, spacing: Spacing.sm) {
                        Text("Elevated Card")
                            .font(.headline)
                        Text("This is an elevated card with shadow")
                            .font(.body)
                            .foregroundColor(Colors().textSecondary)
                    }
                }

                // Outlined card with header
                TchatCard(
                    title: "Card with Header",
                    subtitle: "Subtitle text",
                    leadingIcon: "star.fill",
                    variant: .outlined,
                    isInteractive: true,
                    onTap: { print("Card tapped") }
                ) {
                    VStack(alignment: .leading, spacing: Spacing.xs) {
                        Text("This card has a header and is interactive")
                        Text("Tap to interact")
                            .font(.caption)
                            .foregroundColor(Colors().primary)
                    }
                }

                // Filled card
                TchatCard(variant: .filled, size: .compact) {
                    HStack {
                        Image(systemName: "message.fill")
                            .foregroundColor(Colors().primary)
                        Text("Compact filled card")
                            .font(.subheadline)
                    }
                }

                // Glass card
                TchatCard(variant: .glass, size: .expanded) {
                    VStack(alignment: .leading, spacing: Spacing.md) {
                        TchatCardHeader(
                            title: "Glass Card",
                            subtitle: "Expanded size",
                            leadingIcon: "sparkles"
                        )

                        Text("This is a glass-style card with expanded padding")

                        TchatCardFooter {
                            AnyView(
                                HStack {
                                    TchatButton("Action", variant: .primary, size: .small) {
                                        print("Action tapped")
                                    }
                                    Spacer()
                                    Text("Footer content")
                                        .font(.caption)
                                        .foregroundColor(Colors().textTertiary)
                                }
                            )
                        }
                    }
                }

                // Interactive card with complex content
                TchatCard(
                    variant: .elevated,
                    isInteractive: true,
                    onTap: { print("Complex card tapped") }
                ) {
                    VStack(alignment: .leading, spacing: Spacing.md) {
                        HStack {
                            VStack(alignment: .leading) {
                                Text("Interactive Card")
                                    .font(.title3)
                                    .fontWeight(.semibold)
                                Text("With complex content")
                                    .font(.caption)
                                    .foregroundColor(Colors().textSecondary)
                            }
                            Spacer()
                            Image(systemName: "chevron.right")
                                .foregroundColor(Colors().textTertiary)
                        }

                        HStack {
                            Label("Feature 1", systemImage: "checkmark.circle.fill")
                                .foregroundColor(Colors().success)
                            Spacer()
                            Label("Feature 2", systemImage: "star.fill")
                                .foregroundColor(Colors().warning)
                        }
                        .font(.caption)
                    }
                }
            }
            .padding()
        }
        .background(Color(.systemGroupedBackground))
        .previewLayout(.sizeThatFits)
    }
}
#endif